File: solutions/java/code-carl/ch10_monotone_stack/T5_LC0084_LargestRectangleArea.java
/**
 * Practice template for:
 * - LeetCode 84. Largest Rectangle in Histogram
 *   https://leetcode.cn/problems/largest-rectangle-in-histogram/
 */
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/29 18:59
 * @comment
 */
public class T5_LC0084_LargestRectangleArea {
    public int largestRectangleArea(int[] heights) {
        int[] h = new int[heights.length + 2];
        System.arraycopy(heights, 0, h, 1, heights.length);
        int maxSum = 0;
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < h.length; i++) {
            while (!stack.isEmpty() && h[i] < h[stack.peek()]) {
                int y = h[stack.pop()];
                if (!stack.isEmpty()) {
                    int x = i - stack.peek() - 1;
                    maxSum = Math.max(maxSum, x * y);
                }
            }
            if (!stack.isEmpty() && h[stack.peek()] == h[i]) {
                stack.pop();
            }
            stack.push(i);
        }
        return maxSum;
    }
}

File: solutions/code-carl/ch10_monotone_stack/t5_lc0084_largest_rectangle_area.py
"""LeetCode 84. Largest Rectangle in Histogram
https://leetcode.cn/problems/largest-rectangle-in-histogram/
"""

from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.insert(0, 0)
        heights.append(0)
        stack: List[int] = []
        ans = 0

        for i, curr_h in enumerate(heights):
            while stack and curr_h < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i - stack[-1] - 1
                ans = max(ans, w * h)
            stack.append(i)

        return ans


if __name__ == "__main__":
    solution = Solution()
    assert solution.largestRectangleArea([2, 1, 5, 6, 2, 3]) == 10
    print("All tests passed.")
